x-bytestring/src/X/Data/ByteString.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module X.Data.ByteString (
+    eq1
+  , eq2
+  , eq3
+  , eq4
+  , eq5
+
+  , module Data.ByteString
+  ) where
+
+import           Data.Bits
+import           Data.ByteString
+import           Data.ByteString.Internal (ByteString(..), accursedUnutterablePerformIO)
+import           Data.Word (Word8, Word16, Word32)
+
+import           Foreign.ForeignPtr (withForeignPtr)
+import           Foreign.Storable (peekByteOff)
+
+import           P
+
+
+-- | Fast equality for a 1-byte string.
+eq1 :: Word8 -> ByteString -> Bool
+eq1 x0 (PS fp off len) =
+  let
+    ok =
+      accursedUnutterablePerformIO $! withForeignPtr fp $! \ptr -> do
+        !y0 <- peekByteOff ptr off
+        pure $!
+          x0 == y0
+  in
+    len == 1 && ok
+{-# INLINE eq1 #-}
+
+-- | Fast equality for a 2-byte string.
+eq2 :: Word8 -> Word8 -> ByteString -> Bool
+eq2 x0 x1 (PS fp off len) =
+  let
+    x01 :: Word16
+    !x01 =
+      (fromIntegral x0) .|.
+      (fromIntegral x1 `shiftL` 8)
+
+    ok =
+      accursedUnutterablePerformIO $! withForeignPtr fp $! \ptr -> do
```

Comment:
I haven't benchmarked it, but it's pretty easy to see that there will be a difference by looking at the core:
```hs

--
-- Haskell
--

prop_eq5x :: Property
prop_eq5x =
  gamble arbitrary $ \xs ->
    Char8.eq5 'f' 'a' 'l' 's' 'e' xs

--
-- Core: unsafePerformIO
--

prop_eq5x2 :: ByteString -> Property
prop_eq5x2 =
  \ (x :: ByteString) ->
    counterexample_$scounterexample1
      (ppShow $fShowByteString x)
      (let! { PS dt dt1 dt2 dt3 ~ _ <- x } in
       case dt3 of _ {
         __DEFAULT -> False;
         5# ->
           let! { (# ipv, ipv1 #) ~ _
           <- runRW#
                (\ (s :: State# RealWorld) ->
                   let! { __DEFAULT ~ s' <- noDuplicate# s } in
                   let! { (# ipv, ipv1 #) ~ _
                   <- readWord32OffAddr# (plusAddr# dt dt2) 0# s' } in
                   let! { (# ipv2, ipv3 #) ~ _
                   <- readWord8OffAddr# (plusAddr# dt (+# dt2 4#)) 0# ipv } in
                   let! { __DEFAULT ~ x1
                   <- case ipv1 of _ {
                        __DEFAULT -> False;
                        8348006## ->
                          case ipv3 of _ {
                            __DEFAULT -> False;
                            101## -> True
                          }
                      } } in
                   let! { __DEFAULT ~ s'1 <- touch# dt1 ipv2 } in (# s'1, x1 #)) } in
           ipv1
       })

--
-- Core: unsafeDupablePerformIO
--

prop_eq5x2 :: ByteString -> Property
prop_eq5x2 =
  \ (x :: ByteString) ->
    counterexample_$scounterexample1
      (ppShow $fShowByteString x)
      (let! { PS dt dt1 dt2 dt3 ~ _ <- x } in
       case dt3 of _ {
         __DEFAULT -> False;
         5# ->
           let! { (# ipv, ipv1 #) ~ _
           <- runRW#
                (\ (s :: State# RealWorld) ->
                   let! { (# ipv, ipv1 #) ~ _
                   <- readWord32OffAddr# (plusAddr# dt dt2) 0# s } in
                   let! { (# ipv2, ipv3 #) ~ _
                   <- readWord8OffAddr# (plusAddr# dt (+# dt2 4#)) 0# ipv } in
                   let! { __DEFAULT ~ x1
                   <- case ipv1 of _ {
                        __DEFAULT -> False;
                        8348006## ->
                          case ipv3 of _ {
                            __DEFAULT -> False;
                            101## -> True
                          }
                      } } in
                   let! { __DEFAULT ~ s' <- touch# dt1 ipv2 } in (# s', x1 #)) } in
           ipv1
       })
--
-- Core: accursedUnutterablePerformIO
--

prop_eq5x2 :: ByteString -> Property
prop_eq5x2 =
  \ (x :: ByteString) ->
    counterexample_$scounterexample1
      (ppShow $fShowByteString x)
      (let! { PS dt dt1 dt2 dt3 ~ _ <- x } in
       case dt3 of _ {
         __DEFAULT -> False;
         5# ->
           let! { (# ipv, ipv1 #) ~ _
           <- readWord32OffAddr# (plusAddr# dt dt2) 0# realWorld# } in
           let! { (# ipv2, ipv3 #) ~ _
           <- readWord8OffAddr# (plusAddr# dt (+# dt2 4#)) 0# ipv } in
           let! { __DEFAULT ~ x1
           <- case ipv1 of _ {
                __DEFAULT -> False;
                8348006## ->
                  case ipv3 of _ {
                    __DEFAULT -> False;
                    101## -> True
                  }
              } } in
           let! { __DEFAULT ~ _ <- touch# dt1 ipv2 } in x1
       })
```